[Inductor][Grouped Gemm] Add Improved Blackwell CuTeDSL Kernel

NikhilAPatel opened the pull request:
Summary: Adds Shikai Li's improved GroupGemm into Inductor, replacing the standard version copied from Nvidia's Cutlass repository. This adds device metadata support.

Differential Revision: D88113306


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

drisspg commented on torch/_inductor/kernel/templates/cutedsl_mm_grouped_launcher.py.jinja:
```diff
@@ -0,0 +1,415 @@
+    m_g_i32 = m_end_i32 - m_beg_i32
+
+    a_byte_off = (
+        cutlass.Int64(m_beg_i32) * stride_A_m_elems * cutlass.Int64(sizeof_element)
```
device assert?

drisspg replied:
So the best thing I could come up with is 
```Python
  if handle.index < NUM_STAGES + 100:  # dynamic expression
      cute.printf("ASSERTION FAILED: Index out of bounds: {}", handle.index)
      nvvm.breakpoint()  # poor man device assert
```

this gave me 

``` Shell

ASSERTION FAILED: Index out of bounds: 0
Traceback (most recent call last):
  File "/home/dev/meta/transformer_nuggets/transformer_nuggets/cute/warps.py", line 99, in <module>
    print(inpt)
  File "/home/dev/meta/pytorch/torch/_tensor.py", line 572, in __repr__
    return torch._tensor_str._str(self, tensor_contents=tensor_contents)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dev/meta/pytorch/torch/_tensor_str.py", line 724, in _str
    return _str_intern(self, tensor_contents=tensor_contents)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dev/meta/pytorch/torch/_tensor_str.py", line 643, in _str_intern
    tensor_str = _tensor_str(self, indent)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dev/meta/pytorch/torch/_tensor_str.py", line 375, in _tensor_str
    formatter = _Formatter(get_summarized_data(self) if summarize else self)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dev/meta/pytorch/torch/_tensor_str.py", line 155, in __init__
    tensor_view, torch.isfinite(tensor_view) & tensor_view.ne(0)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
torch.AcceleratorError: CUDA error: unspecified launch failure
Search for `cudaErrorLaunchFailure' in https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html for more information.
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
```

ngimel replied:
This is loads better then getting IMA in the gemm kernel itself. But CuTEDSL doesn't allow you to do first class asserts? That seems like an oversight

drisspg commented on torch/_inductor/kernel/templates/cutedsl_mm_grouped_launcher.py.jinja:
```diff
@@ -0,0 +1,422 @@
+    )
+
+    global _HARDWARE_INFO
+    if _HARDWARE_INFO is None:
```
nit I think its cleaner to `lru_cahce(max_size_0) and get

drisspg commented on torch/_inductor/kernel/templates/cutedsl_mm_grouped_launcher.py.jinja:
```diff
@@ -0,0 +1,422 @@
+        GroupedGemmKernel.bytes_per_tensormap // 8,
+    )
+
+    torch_tensor_of_tensormap = torch.empty(
```
im not sure if anything weird happends if we dont let inductor handle the allocation

NikhilAPatel replied:
@eellison Do you have any insight into this?

NikhilAPatel committed:
[Inductor][Grouped Gemm] Add Improved Blackwell CuTeDSL Kernel (#169336)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/169336

Adds levendlee's improved GroupGemm from D87682597 into Inductor, replacing the standard version copied from Nvidia's Cutlass repository. This adds device metadata support.

Test Plan:
Tritonbench:
`CUDA_VISIBLE_DEVICES=2 TRITON_PRINT_AUTOTUNING=1 TRITON_ALWAYS_COMPILE=1 TORCH_LOGS=+inductor TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 TORCHINDUCTOR_MAX_AUTOTUNE_GEMM=1 python run.py --op grouped_gemm --only aten_grouped_mm,preprocessed_pt2_cute_grouped_mm --precision bf16 --num-inputs 1 --metrics tflops,accuracy`

Unit Tests:
`clear; python test/inductor/test_cutedsl_grouped_mm.py`

Differential Revision: D88113306

alexsamardzic commented:
Some quick benchmarking results with 4 groups used (and admittedly from very limited subset of the (M,N,K) space):

<details>
<summary>Benchmarking script</summary>

```Python
import itertools

import pandas as pd

from tqdm import tqdm

import torch
from triton.testing import do_bench


device = "cuda"
dtype_AB = torch.bfloat16
dtype_offset = torch.int32
dtype_C = torch.bfloat16
ngroups = 4
strided = True

f_ref = torch._grouped_mm
f = torch.compile(
    f_ref,
    options={
        "max_autotune": True,
        "max_autotune_gemm_backends": "CUTEDSL",
       "max_autotune_gemm_search_space": "EXHAUSTIVE",
    },
    dynamic=False,
)


def benchmark_microseconds(f, *args, **kwargs):
    return do_bench(lambda: f(*args, **kwargs), return_mode="median") * 1e3


def get_problem(M, N, K):
    A = torch.randn(M * ngroups, K, device=device).to(dtype_AB)
    B = torch.randn(ngroups, N, K, device=device).to(dtype_AB)
    offs = torch.arange(M, ngroups * M + 1, M, device=device, dtype=dtype_offset)

    return A, B, offs


def benchmark(m: int, k: int, n: int):
    torch._dynamo.reset()

    for i in range(5):
        A, B, offs = get_problem(m, n, k)
        C_ref = f_ref(A, B.transpose(-2, -1), offs, out_dtype=dtype_C)
        C = f(A, B.transpose(-2, -1), offs, out_dtype=dtype_C)
        torch.testing.assert_close(C, C_ref)

    A, B, offs = get_problem(m, n, k)

    eager_time = benchmark_microseconds(
        f_ref,
        A,
        B.transpose(-2, -1),
        offs,
        out_dtype=dtype_C,
    )
    compiled_time = benchmark_microseconds(
        f,
        A,
        B.transpose(-2, -1),
        offs,
        out_dtype=dtype_C,
    )

    return {
        "m": m,
        "n": n,
        "k": k,
        "Eager (CUTLASS) latency (us)": eager_time,
        "Compiled (CuTe) latency (us)": compiled_time,
        "Compiled speedup (d/s)": eager_time / compiled_time,
    }


if __name__ == "__main__":
    k_vals = (2048, 4096, 8192, 16384)
    n_vals = (2048, 4096, 8192, 16384)
    results = []
    i_range = range(6, 14)
    for m in tqdm([1 << i for i in i_range]):
        for n, k in itertools.product(n_vals, k_vals):
            try:
                results.append(benchmark(m, k, n))
            except Exception as e:
                pass

    df = pd.DataFrame(results)
    df.to_csv(f"grouped_mm_2d_3d_time_results_cute.csv", index=False)
    print(df.to_markdown(index=False))
```
</details>

<details>
<summary>Benchmarking results</summary>

|   m |     n |     k |   Eager (CUTLASS) latency (us) |   Compiled (CuTe) latency (us) |   Compiled speedup (d/s) |
|----:|------:|------:|-------------------------------:|-------------------------------:|-------------------------:|
|  64 |  2048 |  2048 |                         32.768 |                         80.992 |                 0.404583 |
|  64 |  2048 |  4096 |                         49.184 |                        135.04  |                 0.364218 |
|  64 |  2048 |  8192 |                         81.952 |                        247.808 |                 0.330708 |
|  64 |  2048 | 16384 |                        139.296 |                        472.064 |                 0.295079 |
|  64 |  4096 |  2048 |                         36.896 |                         81.952 |                 0.450215 |
|  64 |  4096 |  4096 |                         59.392 |                        139.264 |                 0.426471 |
|  64 |  4096 |  8192 |                         92.192 |                        251.904 |                 0.365981 |
|  64 |  4096 | 16384 |                        155.68  |                        481.28  |                 0.323471 |
|  64 |  8192 |  2048 |                         57.376 |                         82.976 |                 0.691477 |
|  64 |  8192 |  4096 |                         92.192 |                        143.36  |                 0.64308  |
|  64 |  8192 |  8192 |                        159.744 |                        262.08  |                 0.609524 |
|  64 |  8192 | 16384 |                        292.992 |                        496.672 |                 0.58991  |
|  64 | 16384 |  2048 |                         94.24  |                         89.664 |                 1.05104  |
|  64 | 16384 |  4096 |                        159.744 |                        155.776 |                 1.02547  |
|  64 | 16384 |  8192 |                        292.864 |                        288.768 |                 1.01418  |
|  64 | 16384 | 16384 |                        560.624 |                        562.208 |                 0.997183 |
| 128 |  2048 |  2048 |                         32.768 |                         84     |                 0.390095 |
| 128 |  2048 |  4096 |                         51.2   |                         83.52  |                 0.613027 |
| 128 |  2048 |  8192 |                         82.032 |                        135.2   |                 0.606746 |
| 128 |  2048 | 16384 |                        139.296 |                        249.888 |                 0.557434 |
| 128 |  4096 |  2048 |                         38.848 |                         82.624 |                 0.470178 |
| 128 |  4096 |  4096 |                         59.392 |                         81.952 |                 0.724717 |
| 128 |  4096 |  8192 |                         94.144 |                        141.312 |                 0.666214 |
| 128 |  4096 | 16384 |                        155.712 |                        258.08  |                 0.603348 |
| 128 |  8192 |  2048 |                         57.376 |                         82.976 |                 0.691477 |
| 128 |  8192 |  4096 |                         92.224 |                         90.016 |                 1.02453  |
| 128 |  8192 |  8192 |                        159.776 |                        153.6   |                 1.04021  |
| 128 |  8192 | 16384 |                        290.912 |                        290.912 |                 1        |
| 128 | 16384 |  2048 |                         94.128 |                         90.144 |                 1.0442   |
| 128 | 16384 |  4096 |                        161.792 |                        158.784 |                 1.01894  |
| 128 | 16384 |  8192 |                        299.04  |                        296.864 |                 1.00733  |
| 128 | 16384 | 16384 |                        575.488 |                        575.52  |                 0.999944 |
| 256 |  2048 |  2048 |                         30.72  |                         80.928 |                 0.379597 |
| 256 |  2048 |  4096 |                         47.104 |                         81.952 |                 0.574776 |
| 256 |  2048 |  8192 |                         73.728 |                         80.928 |                 0.911032 |
| 256 |  2048 | 16384 |                        118.816 |                        139.296 |                 0.852975 |
| 256 |  4096 |  2048 |                         34.816 |                         81.952 |                 0.424834 |
| 256 |  4096 |  4096 |                         53.28  |                         79.904 |                 0.6668   |
| 256 |  4096 |  8192 |                         84     |                         88.192 |                 0.952467 |
| 256 |  4096 | 16384 |                        141.312 |                        161.824 |                 0.873245 |
| 256 |  8192 |  2048 |                         53.28  |                         79.84  |                 0.667335 |
| 256 |  8192 |  4096 |                         84.032 |                         91.168 |                 0.921727 |
| 256 |  8192 |  8192 |                        145.408 |                        161.792 |                 0.898734 |
| 256 |  8192 | 16384 |                        266.304 |                        307.36  |                 0.866424 |
| 256 | 16384 |  2048 |                         83.968 |                         92.16  |                 0.911111 |
| 256 | 16384 |  4096 |                        145.408 |                        168.032 |                 0.865359 |
| 256 | 16384 |  8192 |                        270.496 |                        319.616 |                 0.846316 |
| 256 | 16384 | 16384 |                        520.256 |                        618.496 |                 0.841163 |
| 512 |  2048 |  2048 |                         32.768 |                         82.944 |                 0.395062 |
| 512 |  2048 |  4096 |                         49.152 |                         82.592 |                 0.595118 |
| 512 |  2048 |  8192 |                         77.856 |                         83.008 |                 0.937934 |
| 512 |  2048 | 16384 |                        129.088 |                        151.536 |                 0.851864 |
| 512 |  4096 |  2048 |                         47.104 |                         82.976 |                 0.567682 |
| 512 |  4096 |  4096 |                         75.776 |                         84     |                 0.902095 |
| 512 |  4096 |  8192 |                        129.056 |                        149.6   |                 0.862674 |
| 512 |  4096 | 16384 |                        237.6   |                        290.816 |                 0.817011 |
| 512 |  8192 |  2048 |                         75.776 |                         84     |                 0.902095 |
| 512 |  8192 |  4096 |                        131.04  |                        153.632 |                 0.852947 |
| 512 |  8192 |  8192 |                        243.744 |                        289.792 |                 0.8411   |
| 512 |  8192 | 16384 |                        466.944 |                        559.664 |                 0.834329 |
| 512 | 16384 |  2048 |                        122.88  |                        145.408 |                 0.84507  |
| 512 | 16384 |  4096 |                        226.224 |                        286.784 |                 0.788831 |
| 512 | 16384 |  8192 |                        437.344 |                        548.896 |                 0.79677  |
| 512 | 16384 | 16384 |                        850.992 |                       1151.94  |                 0.738749 |
| 1024 |  2048 |  2048 |                         45.056 |                         85.024 |                 0.529921 |
| 1024 |  2048 |  4096 |                         69.632 |                         83.968 |                 0.829268 |
| 1024 |  2048 |  8192 |                        120.864 |                        145.408 |                 0.831206 |
| 1024 |  2048 | 16384 |                        226.336 |                        288.784 |                 0.783755 |
| 1024 |  4096 |  2048 |                         69.632 |                         84     |                 0.828952 |
| 1024 |  4096 |  4096 |                        118.88  |                        145.408 |                 0.817562 |
| 1024 |  4096 |  8192 |                        223.232 |                        274.528 |                 0.813148 |
| 1024 |  4096 | 16384 |                        429.984 |                        552.96  |                 0.777604 |
| 1024 |  8192 |  2048 |                        112.736 |                        137.248 |                 0.821404 |
| 1024 |  8192 |  4096 |                        209.936 |                        262.08  |                 0.801038 |
| 1024 |  8192 |  8192 |                        407.616 |                        514.176 |                 0.792756 |
| 1024 |  8192 | 16384 |                        820.192 |                        984.096 |                 0.833447 |
| 1024 | 16384 |  2048 |                        217.088 |                        266.208 |                 0.815483 |
| 1024 | 16384 |  4096 |                        405.504 |                        509.92  |                 0.795231 |
| 1024 | 16384 |  8192 |                        832.608 |                       1063.97  |                 0.78255  |
| 1024 | 16384 | 16384 |                       1551.52  |                       1930.24  |                 0.803796 |
| 2048 |  2048 |  2048 |                         67.584 |                         80.896 |                 0.835443 |
| 2048 |  2048 |  4096 |                        116.768 |                        139.264 |                 0.838465 |
| 2048 |  2048 |  8192 |                        223.232 |                        280.544 |                 0.795711 |
| 2048 |  2048 | 16384 |                        405.536 |                        555.008 |                 0.730685 |
| 2048 |  4096 |  2048 |                        108.544 |                        131.104 |                 0.827923 |
| 2048 |  4096 |  4096 |                        202.72  |                        248.768 |                 0.814896 |
| 2048 |  4096 |  8192 |                        392.256 |                        520.096 |                 0.754199 |
| 2048 |  4096 | 16384 |                        765.984 |                       1019.41  |                 0.751401 |
| 2048 |  8192 |  2048 |                        208.896 |                        249.952 |                 0.835744 |
| 2048 |  8192 |  4096 |                        389.184 |                        474.128 |                 0.820842 |
| 2048 |  8192 |  8192 |                        767.984 |                        999.424 |                 0.768427 |
| 2048 |  8192 | 16384 |                       1573.89  |                       1932.32  |                 0.814507 |
| 2048 | 16384 |  2048 |                        390.16  |                        479.264 |                 0.814082 |
| 2048 | 16384 |  4096 |                        778.24  |                       1019.9   |                 0.763052 |
| 2048 | 16384 |  8192 |                       1558.56  |                       1992.69  |                 0.78214  |
| 2048 | 16384 | 16384 |                       3102.72  |                       3696.67  |                 0.839328 |
| 4096 |  2048 |  2048 |                        102.4   |                        128.896 |                 0.794439 |
| 4096 |  2048 |  4096 |                        197.632 |                        249.76  |                 0.791288 |
| 4096 |  2048 |  8192 |                        399.392 |                        503.904 |                 0.792595 |
| 4096 |  2048 | 16384 |                        767.024 |                        974.752 |                 0.786891 |
| 4096 |  4096 |  2048 |                        204.768 |                        247.808 |                 0.826317 |
| 4096 |  4096 |  4096 |                        393.216 |                        474.096 |                 0.829402 |
| 4096 |  4096 |  8192 |                        790.624 |                       1061.38  |                 0.744905 |
| 4096 |  4096 | 16384 |                       1536.06  |                       1959.95  |                 0.783725 |
| 4096 |  8192 |  2048 |                        379.872 |                        482.304 |                 0.787619 |
| 4096 |  8192 |  4096 |                        761.92  |                        905.28  |                 0.84164  |
| 4096 |  8192 |  8192 |                       1553.39  |                       1897.44  |                 0.818678 |
| 4096 |  8192 | 16384 |                       3120.16  |                       3749.98  |                 0.832046 |
| 4096 | 16384 |  2048 |                        757.776 |                        914.368 |                 0.828743 |
| 4096 | 16384 |  4096 |                       1516.51  |                       1874.98  |                 0.808817 |
| 4096 | 16384 |  8192 |                       3085.25  |                       3781.63  |                 0.815851 |
| 4096 | 16384 | 16384 |                       6261.81  |                       7458.88  |                 0.83951  |
</details>

For some reason, the compiled version of the operator throws an exception when M changed, so the results above are actually concatenated results of the script runs for individual M values.

I was not able to run the test `pytest -k test_grouped_gemm_compiled_op_2d/3d test/test_matmul_cuda.py` (with `"max_autotune_gemm_backends": "CUTEDSL"`, and `dynamic=False` added), because `False` was returned from `can_use_tma` (while the Triton auto-tuned kernel handles these inputs fine).

Some additional benchmarking results, with my Gluon kernel from #166063 thrown into the mix, are [here](https://github.com/pytorch/pytorch/pull/166063#issuecomment-3623049528).  Note that the Gluon kernel is tuned over multitude of configs (also it uses a triton-lang/triton#8786 that it's uncertain will it be accepted), so it has an advantage in that regards.  On the other hand, Triton/Gluon do not have 2-CTA support yet (it's in the works), and I believe this is something that could help its performance further.  In any case, it seems all of our compiled version of kernels need more work to come up to the level of eager-mode CUTLASS C++ version...

alexsamardzic commented:
No hurry 🙂
